DS_and_Algorithm/Selenium/test.java

```diff
@@ -0,0 +1,706 @@
+    driver.get(pages.formPage);
+
+    WebElement originallySelected = enabledSelectedRadioButton();
+    assertSelected(originallySelected);
+
+    WebElement toSelect = enabledUnselectedRadioButton();
+    assertNotSelected(toSelect);
+
+    toSelect.click();
+    assertNotSelected(originallySelected);
+    assertSelected(toSelect);
+  }
+
+  @Test
+  void testShouldBeAbleToToggleAnEnabledUnselectedCheckbox() {
+    driver.get(pages.formPage);
+    assertCanToggle(enabledUnselectedCheckbox());
+  }
+
+  @Test
+  void testShouldBeAbleToToggleAnEnabledSelectedCheckbox() {
+    driver.get(pages.formPage);
+    assertCanToggle(enabledSelectedCheckbox());
+  }
+
+  @Test
+  void testClickingOnASelectedRadioButtonShouldLeaveItSelected() {
+    driver.get(pages.formPage);
+
+    WebElement button = enabledSelectedRadioButton();
+    assertThat(button.isSelected()).isTrue();
+
+    button.click();
+
+    assertThat(button.isSelected()).isTrue();
+  }
+
+  @Test
+  void testShouldBeAbleToToggleEnabledMultiSelectOption() {
+    driver.get(pages.formPage);
+    assertCanToggle(selectedMultipleSelectOption());
+  }
+
+  @Test
+  void testShouldBeAbleToToggleSelectableCheckboxByClickingOnIt() {
+    driver.get(pages.formPage);
+
+    WebElement checkbox = enabledUnselectedCheckbox();
+    assertNotSelected(checkbox);
+
+    checkbox.click();
+    assertSelected(checkbox);
+
+    checkbox.click();
+    assertNotSelected(checkbox);
+  }
+
+  @Test
+  void testShouldBeAbleToSelectSelectableRadioButtonByClickingOnIt() {
+    driver.get(pages.formPage);
+
+    WebElement radioButton = enabledUnselectedRadioButton();
+    assertNotSelected(radioButton);
+
+    radioButton.click();
+    assertSelected(radioButton);
+
+    radioButton.click();
+    assertSelected(radioButton);
+  }
+
+  @Test
+  void testClickingDisabledSelectedCheckboxShouldBeNoop() {
+    driver.get(pages.formPage);
+    assertClickingPreservesCurrentlySelectedStatus(randomlyDisabledSelectedCheckbox());
+  }
+
+  @Test
+  void testClickingDisabledUnselectedCheckboxShouldBeNoop() {
+    driver.get(pages.formPage);
+    assertClickingPreservesCurrentlySelectedStatus(disabledUnselectedCheckbox());
+  }
+
+  @Test
+  void testClickingDisabledSelectedRadioButtonShouldBeNoop() {
+    driver.get(pages.formPage);
+    assertClickingPreservesCurrentlySelectedStatus(disabledSelectedRadioButton());
+  }
+
+  @Test
+  void testClickingDisabledUnselectedRadioButtonShouldBeNoop() {
+    driver.get(pages.formPage);
+    assertClickingPreservesCurrentlySelectedStatus(disabledUnselectedRadioButton());
+  }
+
+  private void assertNotSelected(WebElement element) {
+    assertSelected(element, UNSELECTED);
+  }
+
+  private void assertSelected(WebElement element) {
+    assertSelected(element, SELECTED);
+  }
+
+  private void assertSelected(WebElement element, boolean isSelected) {
+    wait.until(ExpectedConditions.elementSelectionStateToBe(element, isSelected));
```

Comment:
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/Amruta101998/DS-and-Algorithm/pull/16#issuecomment-2211163237>#b402cc</a><span> - 07/05/2024, 04:55 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The use of 'wait.until(ExpectedConditions.elementSelectionStateToBe(element, isSelected));' can lead to performance issues due to potential long waits if the condition is not met promptly.
</div>
<div id="fix">
<b>Fix:</b> Consider using a more efficient polling mechanism or reducing the wait time to minimize the performance impact.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -390,7 +390,7 @@
     wait.until(ExpectedConditions.elementSelectionStateToBe(element, isSelected));
 +   wait.withTimeout(Duration.ofSeconds(5)).pollingEvery(Duration.ofMillis(500)).until(ExpectedConditions.elementSelectionStateToBe(element, isSelected));
```
</div>
</details>
</div>
<br>


</div>